builder/eth_service.go

```diff
@@ -0,0 +1,85 @@
+package builder
+
+import (
+	"errors"
+	"time"
+
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/miner"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type IEthereumService interface {
+	BuildBlock(attrs *BuilderPayloadAttributes) (*engine.ExecutionPayloadEnvelope, error)
+	GetBlockByHash(hash common.Hash) *types.Block
+	Config() *params.ChainConfig
+	Synced() bool
+}
+
+type EthereumService struct {
+	eth *eth.Ethereum
+}
+
+func NewEthereumService(eth *eth.Ethereum) *EthereumService {
+	return &EthereumService{
+		eth: eth,
+	}
+}
+
+func (s *EthereumService) BuildBlock(attrs *BuilderPayloadAttributes) (*engine.ExecutionPayloadEnvelope, error) {
+	// Send a request to generate a full block in the background.
+	// The result can be obtained via the returned channel.
+	args := &miner.BuildPayloadArgs{
+		Parent:       attrs.HeadHash,
+		Timestamp:    uint64(attrs.Timestamp),
+		FeeRecipient: attrs.SuggestedFeeRecipient, // TODO (builder): use builder key as fee recipient
+		GasLimit:     &attrs.GasLimit,
+		Random:       attrs.Random,
+		Withdrawals:  attrs.Withdrawals,
+		BeaconRoot:   attrs.ParentBeaconBlockRoot,
+		Transactions: attrs.Transactions,
+		NoTxPool:     attrs.NoTxPool,
+	}
+
+	payload, err := s.eth.Miner().BuildPayload(args)
+	if err != nil {
+		log.Error("Failed to build payload", "err", err)
+		return nil, err
+	}
+
+	resCh := make(chan *engine.ExecutionPayloadEnvelope, 1)
+	go func() {
+		resCh <- payload.ResolveFull()
+	}()
+
+	timer := time.NewTimer(2 * time.Second)
```

Comment:
I see that it's set to 4 seconds for L1 builder. should it be dependent on block time? if block building takes longer than 2 seconds it's already a failure

Reply:
yeah using block time as default for now, can update as additional param later